src/utils/yup/tests/index.ts

```diff
@@ -0,0 +1,219 @@
+  createError: TestContext['createError'],
+): Promise<Q | ValidationError> {
+  try {
+    const { data } = await apolloClient.query<Q, V>({
+      query: queryDocument,
+      variables,
+    });
+    return data;
+  } catch (e) {
+    if (e.message.includes('Failed to fetch')) {
+      return createError({
+        message: formatMessage({ id: 'error.network' }),
+      });
+    }
+
+    console.error(e);
+    return createUnknownError(createError);
+  }
+}
+
+export interface CustomTestConfig {
+  query: DocumentNode;
+  isOptional?: boolean;
+  circuitBreaker?: boolean | ((value) => boolean);
+  debounceInput?: boolean;
+  wait?: number;
+}
+
+/**
+ * Generate a custom test function for yup .test.
+ * Expects query to take exactly one variable, namely, the value of the input field.
+ * @param query Query to be executed
+ * @param isOptional Is field required? False by default
+ * @param circuitBreaker False by default. If true, will not run test. Can take a test function from a previous test. In this case,
+ * will not run if the function resolves to false. Previous tests should therefore assert positively, e.g. isUsernameValid. If it
+ * is not valid, do not run this test...
+ * @returns yup TestFunction, that runs the provided query using the input value
+ * as the query's input variable. Function will return true if there are no results (i.e. value doesn't exist in db);
+ * false if there are (value does exist in db); and an appropriate error message if the query errors.
+ */
+export function createYupTestFromQuery({
+  query,
+  isOptional = false,
+  circuitBreaker = false,
+  debounceInput = true,
+  wait = 200,
+}: CustomTestConfig) {
+  if (debounceInput) {
+    return yupDebounce(checkIfValueExistsInDB, wait, {
+      isOptional,
+      circuitBreaker,
+    });
+  }
+
+  return checkIfValueExistsInDB;
+
+  async function checkIfValueExistsInDB(value) {
+    const { createError } = this as TestContext;
+
+    const cancel = cancelEarly(circuitBreaker, value);
+
+    if (!value || cancel) {
+      return isOptional;
+    }
+
+    const queryName = (query.definitions[0] as OperationDefinitionNode).name
+      ?.value;
+    const variableKey = (query.definitions[0] as OperationDefinitionNode)
+      .variableDefinitions?.[0].variable.name.value;
+
+    if (!queryName || !variableKey) {
+      throw new Error(
+        'Query must be a named query and have exactly one named variable.',
+      );
+    }
+
+    const result = await runQuery(
+      query,
+      {
+        [variableKey]: value,
+      },
+      createError,
+    );
+
+    if (result instanceof ValidationError) {
+      return result;
+    }
+
+    return !(result as object)[cleanQueryName(queryName)]?.items.length;
+  }
+}
+
+interface YupDebounceOptions {
+  isOptional: NonNullable<CustomTestConfig['isOptional']>;
+  circuitBreaker: NonNullable<CustomTestConfig['circuitBreaker']>;
+}
+/**
+ * Debounce yup validation test function, inspired by lodash's debounce function.
```

Comment:
Added a custom yup validation debounce function. Lodash's debounce doesn't work for this use case as explained below.

Reply:
Thanks for leaving that detailed comment explaining the phenomenon